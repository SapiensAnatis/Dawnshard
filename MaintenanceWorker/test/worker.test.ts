import { Miniflare } from "miniflare";
import { decode, decodeAsync } from "@msgpack/msgpack";

const mf = new Miniflare({
  scriptPath: "./build/worker/shim.mjs",
  modules: true,
  modulesRules: [
    {
      type: "CompiledWasm",
      include: ["**/*.wasm"],
      fallthrough: true,
    },
  ],
  bindings: {
    MAINTENANCE_END_DATE: "2024-05-20T22:00:00+01:00",
    MAINTENANCE_TITLE: "Maintenance",
    MAINTENANCE_BODY:
      "Dawnshard is currently under maintenance so\ndreadfullydistinct can fix his\nConsul install.",
  },
});

describe("web", () => {
  test("renders webpage", async () => {
    const res = await mf.dispatchFetch("http://localhost");

    expect(res.ok).toBe(true);
    expect(await res.text()).toMatchSnapshot();
  });

  test("renders webpage on any other path", async () => {
    const res = await mf.dispatchFetch("http://localhost/aaaaaa");

    expect(res.ok).toBe(true);
    expect(await res.text()).toMatch(/^<!DOCTYPE html>/);
  });
});

describe("game", () => {
  const deserialize = async (blob: Blob) => {
    if (blob.stream) {
      return await decodeAsync(blob.stream());
    } else {
      return decode(await blob.arrayBuffer());
    }
  };

  test("sends maintenance response", async () => {
    const res = await mf.dispatchFetch(
      "http://localhost/2.19.0_20220714193707/dungeon_start/start",
      {
        method: "POST",
      }
    );

    expect(res.ok).toBe(true);
    const decoded = await deserialize(await res.blob());

    expect(decoded).toEqual({
      data_headers: {
        result_code: 101,
      },
      data: {
        result_code: 101,
      },
    });
  });

  test("sends maintenance text", async () => {
    const res = await mf.dispatchFetch(
      "http://localhost/2.19.0_20220719103923/maintenance/get_text",
      {
        method: "POST",
      }
    );

    expect(res.ok).toBe(true);
    const decoded = await deserialize(await res.blob());

    expect(decoded).toEqual({
      data_headers: {
        result_code: 1,
      },
      data: {
        maintenance_text: `<title>Maintenance</title>
<body>Dawnshard is currently under maintenance so
dreadfullydistinct can fix his
Consul install.</body>
<schedule>Check back at:</schedule>
<date>2024-05-21T06:00:00</date>`,
      },
    });
  });
});
