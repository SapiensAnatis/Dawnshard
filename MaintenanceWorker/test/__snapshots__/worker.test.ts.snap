// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`web > renders webpage 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <title>Dawnshard</title>
  <style>
    h1 {
      font-weight: bolder;
      font-size: 3rem;
    }

    h2 {
      font-weight: bold;
      font-size: 2rem;
    }

    p {
      font-size: 1.25rem;
      margin: 0;
    }

    .end-time {
      height: 300px;
    }

    .emoji {
      font-size: 7.5rem;
      margin: 0;
    }

    html,
    body {
      margin: 0;
      height: 100%;
    }

    .message {
      display: flex;
      align-items: center;
      justify-content: center;
      flex-direction: column;
    }

    body {
      font-family: sans-serif;
      color: whitesmoke;
      background-color: #1a264c;
      display: flex;
      align-items: center;
      justify-content: center;
      text-align: center;
    }

    br {
      margin: 1rem auto;
    }

    .loading #date {
      visibility: hidden;
    }
  </style>
  <script>
    document.documentElement.className += "loading";
    window.onload = () => {
      localizeTime();
      document.documentElement.className = document.documentElement.className.replace("loading", "");
    };

    function localizeTime() {
      const dateElement = document.getElementById("date");
      if (!dateElement) {
        return;
      }

      const timestamp = dateElement.getAttribute("data-timestamp");
      if (!timestamp) {
        return;
      }

      const dateObj = new Date(timestamp * 1000);
      document.getElementById("date").textContent = dateObj.toLocaleString(undefined, {
        dateStyle: "medium",
        timeStyle: "short"
      });
    }
  </script>
</head>

<body>
  <div class="message">
    <span class="emoji">ðŸš§</span>
    <h1 class="title">Maintenance</h1>
    <p class="text">Dawnshard is currently under maintenance so
dreadfullydistinct can fix his
Consul install.</p>
    <br />
    <div class="end-time">
      <h2>Expected end time:</h2>
      <p id="date" class="text" data-timestamp="1716238800">20/05/2024 22:00:00 +01:00</p>
    </div>
  </div>
</body>

</html>"
`;
